chroma-agent/chroma_agent/device_plugins/linux_components/block_devices.py

```diff
@@ -4,98 +4,197 @@
-    DISKBYPATHPATH = os.path.join('/dev', 'disk', 'by-path')
-    SYSBLOCKPATH = os.path.join('/sys', 'block')
-    MDRAIDPATH = os.path.join('/dev', 'md')
-    DEVPATH = '/dev'
-    MAXRETRIES = 5
-    non_existent_paths = set([])
-    previous_path_status = {}
+PRECEDENCE = [
+    MAPPER_PATH,
+    DISK_BY_ID_PATH,
+    DISK_BY_PATH_PATH,
+    re.compile('.+')
+]
+
+
+def flat_map(fn, xs):
+    return chain.from_iterable(imap(fn, xs))
+
+
+def get_idx(x):
+    return [index for index, v in enumerate(PRECEDENCE) if v.match(x)][0]
+
+
+def compare(x, y):
+    idx1 = get_idx(x)
+    idx2 = get_idx(y)
+
+    if idx1 == idx2:
+        return 0
+    elif idx1 > idx2:
+        return 1
+
+    return -1
+
+
+def sort_paths(xs):
+    return sorted(xs, cmp=compare)
+
+
+def scanner_cmd(cmd):
+    client = socket.socket(socket.AF_UNIX)
+    client.settimeout(1)
+    client.connect_ex("/var/run/device-scanner.sock")
+    client.sendall(json.dumps({"ACTION": cmd}))
+    client.shutdown(socket.SHUT_WR)
+
+    out = ''
+
+    while True:
+        data = client.recv(1024)
+        size = len(data)
+
+        if size == 0:
+            break
+
+        out += data
+
+    return json.loads(out)
+
+
+def get_major_minor(x):
+    return "%s:%s" % (x.get('MAJOR'), x.get('MINOR'))
 
-    def __init__(self):
-        self.old_udev = None
-        self._major_minor_to_fstype = {}  # Build this map to retrieve fstype in _device_node
 
-        for blkid_dev in utils.BlkId().itervalues():
-            major_minor = self._dev_major_minor(blkid_dev['path'])
+def as_device(x):
+    paths = sort_paths(x.get('PATHS', []))
+    path = next(iter(paths), None)
 
-            if major_minor:
-                self._major_minor_to_fstype[major_minor] = blkid_dev['type']
+    return {
+        'major_minor': get_major_minor(x),
+        'path': path,
+        'paths': paths,
+        'serial_80': x.get('IML_SCSI_80'),
+        'serial_83': x.get('IML_SCSI_83'),
+        'size': int(x['IML_SIZE']) * 512,
+        'filesystem_type': x.get('ID_FS_TYPE'),
+        'device_type': x.get('DEVTYPE'),
+        'device_path': x.get('DEVPATH'),
+        'partition_number': x.get('ID_PART_ENTRY_NUMBER'),
+        'parent': None
+    }
 
-        self.block_device_nodes, self.node_block_devices = self._parse_sys_block()
 
-    def _dev_major_minor(self, path):
-        """ Return a string if 'path' is a block device or link to one, else return None """
+def get_parent_path(p):
+    return os.sep.join(p.split(os.sep)[0:-1])
 
-        file_status = None
-        retries = self.MAXRETRIES
-        while retries > 0:
-            try:
-                file_status = os.stat(path)
 
-                if path in self.non_existent_paths:
-                    self.non_existent_paths.discard(path)
-                    daemon_log.debug('New device started to respond %s' % path)
+def find_device_by_device_path(p, xs):
+    return next(iter([d for d in xs if d['device_path'] == p]), None)
```

Comment:
this can be written: next((d for d in xs if d['device_path'] == p), None) to use generator